### PR TITLE
CON-1082: Publish to new public Artifactory

### PR DESCRIPTION
**Pull request by filipesoliveira**
There are now two public repositories named conclave-maven-dev for SNAPSHOTs and RCs, and conclave-maven-release for GA.
This commit updates the repositories used by Gradle to publish the artifacts.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -111,6 +111,38 @@ static Map<String, String> getEnvironmentVariables(String... envNames) {
     return envNames.collectEntries {envName -> [(envName): System.getenv(envName)] }
 }
 
+String getMavenRepositoryURL() {
+    def url = "${environment.OBLIVIUM_MAVEN_URL}/${environment.OBLIVIUM_MAVEN_REPOSITORY}"
```
Comment (shamsasari) [inline build.gradle]: Hardcode the public repo URLs, as external developers will need to know them anyway.

File: build.gradle [hunk 2]
```diff
@@ -111,6 +111,38 @@ static Map<String, String> getEnvironmentVariables(String... envNames) {
     return envNames.collectEntries {envName -> [(envName): System.getenv(envName)] }
 }
 
+String getMavenRepositoryURL() {
```
Comment (shamsasari) [inline build.gradle]: `getArtifactoryForPublication`

File: build.gradle [hunk 3]
```diff
@@ -111,6 +111,38 @@ static Map<String, String> getEnvironmentVariables(String... envNames) {
...
+    def url = "${environment.OBLIVIUM_MAVEN_URL}/${environment.OBLIVIUM_MAVEN_REPOSITORY}"
+    if (conclave_version.endsWith("-SNAPSHOT")) {
+        return url + "-dev"
+    }
+
+    // Make sure this is a release branch
+    def gitBranchName = getGitBranchName()
+    if(!gitBranchName.startsWith("release/")) {
+        throw new GradleException("Cannot determine which repository to use for publishing the artifact")
+    }
```
Comment (shamsasari) [inline build.gradle]: Not necessary

File: build.gradle [hunk 4]
```diff
@@ -111,6 +111,38 @@ static Map<String, String> getEnvironmentVariables(String... envNames) {
     return envNames.collectEntries {envName -> [(envName): System.getenv(envName)] }
 }
 
+String getMavenRepositoryURL() {
+    def url = "${environment.OBLIVIUM_MAVEN_URL}/${environment.OBLIVIUM_MAVEN_REPOSITORY}"
+    if (conclave_version.endsWith("-SNAPSHOT")) {
```
Comment (shamsasari) [inline build.gradle]: Use the version string for conclave graal, not of Conclave. Yes, for this check they are equivalent, but it's clearer to use the version of the project that is going to be published.

File: scripts/devenv_envs.sh [hunk 5]
```diff
@@ -5,6 +5,6 @@ export OBLIVIUM_CONTAINER_REGISTRY_URL=${OBLIVIUM_CONTAINER_REGISTRY_URL:-"local
 export OBLIVIUM_CONTAINER_REGISTRY_USERNAME=${OBLIVIUM_CONTAINER_REGISTRY_USERNAME:-}
 export OBLIVIUM_CONTAINER_REGISTRY_PASSWORD=${OBLIVIUM_CONTAINER_REGISTRY_PASSWORD:-}
 export OBLIVIUM_MAVEN_URL=${OBLIVIUM_MAVEN_URL:-"https://software.r3.com/artifactory"}
-export OBLIVIUM_MAVEN_REPOSITORY=${OBLIVIUM_MAVEN_REPOSITORY:-"sgxjvm-maven-nightly"}
+export OBLIVIUM_MAVEN_REPOSITORY=${OBLIVIUM_MAVEN_REPOSITORY:-"conclave-maven"}
```
Comment (shamsasari) [inline scripts/devenv_envs.sh]: See comment above. Delete these two env variables.

**Review by shamsasari — CHANGES_REQUESTED** [review]

**Comment by shamsasari** [comment]
@bonifazi-r3 you should probably look at this as well, since you're doing something similar.

File: graal-sdk/build.gradle [hunk 6]
```diff
@@ -44,9 +44,18 @@ publishing {
                 // Details about the artifact being published
                 groupId = "com.r3.conclave"
                 artifactId = "graal-sdk"
-                version = "$graal_version-$conclave_version"
+                version = getConclaveGraalVersion()
```
Comment (shamsasari) [inline graal-sdk/build.gradle:47]: Can we use the dedicated Project `version` field that Gradle already provides.

**Comment by filipesoliveira** [comment]
What should be done about the license for graal-vm?

File: graal-sdk/build.gradle [hunk 7]
```diff
@@ -44,9 +44,18 @@ publishing {
...
                 description = 'Graal SDK'
             }
+
+            pom {
+                licenses {
+                    license {
+                        name = 'Universal Permissive License, Version 1.0'
+                        url = 'http://opensource.org/licenses/UPL'
```
Comment (bon000) [inline graal-sdk/build.gradle:55]: Not sure I understand, what is this license? Isn't the original Graal provided with a GPL + Classpath exception?
If I understand correctly, we only can use 2 licenses
1 - full GPL
2 - GPL with classpath exception

File: build.gradle [hunk 8]
```diff
@@ -111,6 +109,20 @@ static Map<String, String> getEnvironmentVariables(String... envNames) {
     return envNames.collectEntries {envName -> [(envName): System.getenv(envName)] }
 }
 
+String getArtifactoryForPublication() {
+    def url = "https://software.r3.com/artifactory/conclave-maven"
+    def conclaveGraal = getConclaveGraalVersion()
+    if (conclaveGraal.endsWith("-SNAPSHOT")) {
```
Comment (bon000) [inline build.gradle:115]: Few things here:
- I guess we should somehow agree on some formatting in the code, i.e. I would personally put a line before the if and remove it after the code below. Not a blocker though.
- I would make the condition below in fewer lines, not a blocker, but nicer to see
```
def conclaveGraal = getConclaveGraalVersion()
def repositoryPath = if (conclaveGraal.endsWith("-SNAPSHOT")) ? "conclave-maven-dev" : "conclave-maven-release"
return "https://software.r3.com/artifactory/" + repositoryPath
```
Or similar

**Review by bon000 — CHANGES_REQUESTED** [review]

File: graal-sdk/build.gradle (hunk 7 above)
Comment (shamsasari) [reply graal-sdk/build.gradle:55]: GraalVM has different licenses each its components. Have a look at its README: https://github.com/oracle/graal#license.

This artifact is UPL, as evidenced by the published artifact from Oracle: https://repo1.maven.org/maven2/org/graalvm/sdk/graal-sdk/22.2.0/graal-sdk-22.2.0.pom

Comment (shamsasari) [reply graal-sdk/build.gradle:55]: @filipesoliveira this question from Marco shows that we need to have a comment here.

**Comment by shamsasari** [comment]
Just noticed that you were referencing the wrong ticket, updated the PR.

**Comment by shamsasari** [comment]
>What should be done about the license for graal-vm?

Follow my suggestion in the comments for the JIRA ticket, but don't put the ticket into Done once it's merged. Mark it as Blocked so that we can remember to chase up legal for the final answer.

File: graal-sdk/build.gradle (hunk 7 above)
Comment (bon000) [reply graal-sdk/build.gradle:55]: Interesting thanks, so it looks fine. It seems that that license WAS actually GPL+CPE and then they changed it https://github.com/oracle/graal/commit/c90742063beaeed879a29b65eca6abe3930146a6
So with "Graal SDK" I guess they mean "that bunch of configs that assemble things together"?

Comment (shamsasari) [reply graal-sdk/build.gradle:55]: No, The Graal SDK is the Java API to use GraalVM features from the java code that will be compiled with GraalVM. Confusing, I know!

Comment (bon000) [reply graal-sdk/build.gradle:55]: Yep, ok, did not gig further. Sounds interesting anyway how they think that the GPL+CPE does not infect this layer.

**Comment by filipesoliveira** [comment]
I have created a new PR to so the branch looks used is correct.
PR: https://github.com/R3Conclave/graal-patches/pull/5

**Pull request closed without merging**